synapse/lib/view.py

```diff
@@ -138,12 +119,263 @@ async def __anit__(self, core, node):
+
+    async def delMergeVote(self, useriden):
+        return await self._push('merge:vote:del', useriden, s_common.now())
+
+    @s_nexus.Pusher.onPush('merge:vote:del')
+    async def _delMergeVote(self, useriden, tick):
+
+        self.reqParentQuorum()
+        uidn = s_common.uhex(useriden)
+
+        byts = self.core.slab.pop(self.bidn + b'merge:vote' + uidn, db='view:meta')
+
+        await self.tryToMerge(tick)
+
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    async def initMergeTask(self):
+
+        if not self.merging:
+            return
+
+        if not await self.core.isCellActive():
+            return
+
+        self.mergetask = self.core.schedCoro(self.runViewMerge())
+
+    async def finiMergeTask(self):
+        if self.mergetask is not None:
+            self.mergetask.cancel()
+            self.mergetask = None
+
+    async def runViewMerge(self):
+        # run a view merge which eventually results in removing the view and top layer
+        # this routine must be able to be resumed and may assume that the top layer is
+        # not receiving any edits.
+        try:
+
+            # ensure there are none marked dirty
+            await self.layers[0]._saveDirtySodes()
+
+            merge = self.getMergeRequest()
+
+            # merge edits as the merge request user
+            meta = {
+                'user': merge.get('creator'),
+                'merge': merge.get('iden'),
+            }
+
+            async def chunked():
+                nodeedits = []
+
+                async for nodeedit in self.layers[0].iterLayerNodeEdits():
+
+                    nodeedits.append(nodeedit)
+
+                    if len(nodeedits) == 10:
+                        yield nodeedits
+                        nodeedits.clear()
+
+                if nodeedits:
+                    yield nodeedits
+
+            total = self.layers[0].getStorNodeCount()
+
+            count = 0
+            nextprog = 1000
+
+            await self.core.feedBeholder('view:merge:prog', {'view': self.iden, 'count': count, 'total': total})
+
+            async with await self.parent.snap(user=self.core.auth.rootuser) as snap:
+
+                async for edits in chunked():
+
+                    meta['time'] = s_common.now()
+
+                    await snap.saveNodeEdits(edits, meta)
+                    await asyncio.sleep(0)
+
+                    count += len(edits)
+
+                    if count >= nextprog:
+                        await self.core.feedBeholder('view:merge:prog', {'view': self.iden, 'count': count, 'total': total})
+                        nextprog += 1000
+
+            await self.core.feedBeholder('view:merge:fini', {'view': self.iden})
+
+            # remove the view and top layer
+            await self.core.delView(self.iden)
+            await self.core.delLayer(self.layers[0].iden)
+
+        except Exception as e:
+            logger.exception(f'Error while merging view: {self.iden}')
```

Comment:
WE discussed in chat but leaving this here as a reminder:

If you fork view your after a merge request has been created but before it's been approved, then approve the MR, it merges the data but fails to delete the view because it has children. it's a weird edge case but maybe we should prevent forking from a view with an open MR?

Reply:
This is fixed in the current version. :+1: